src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,148 @@
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> previous = getNodeByIndex(index - 1);
+            newNode.next = previous.next;
+            previous.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexCheck(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.element;
+        node.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        return unlink(nodeToRemove);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        int index = 0;
+        do {
+            if (object == current.element
+                    || (current.element != null
+                    && current.element.equals(object))) {
+                remove(index);
+                return true;
+            }
+            index++;
+            current = current.next;
+        } while (current != null);
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private class Node<T> {
+        private T element;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(T element) {
+            this.element = element;
+        }
+    }
+
+    private Node<T> findPreviousNode(Node<T> node) {
+        Node<T> current = head;
+        while (current != null && current.next != node) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private T unlink(Node<T> nodeToRemove) {
+        T removedElement = nodeToRemove.element;
+        if (nodeToRemove == head) {
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> previous = findPreviousNode(nodeToRemove);
+            previous.next = nodeToRemove.next;
+            if (nodeToRemove == tail) {
+                tail = previous;
+            }
+        }
+        size--;
+        return removedElement;
+    }
+
+    private void indexCheck(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Unable to set index: " + index);
```

Comment:
Code duplication try to unite indexCheck and if statement in getNodeByIndex method